docs: rewrite pull query docs page (#7532)
confluentinc/ksql: docs/developer-guide/ksqldb-reference/select-pull-query.md

by using a [CREATE TABLE](../../ksqldb-reference/create-table) statement.

WHERE Clause Guidelines
-----------------------

By default, only key lookups are enabled. They have the following requirements:
-   Key column(s) must use an equality comparison to a literal (e.g. KEY = 'abc').
-   On windowed tables, WINDOWSTART and WINDOWEND can be optionally compared to literals. 
    For more information on windowed tables, see [Time and Windows in ksqlDB](../../concepts/time-and-windows-in-ksqldb-queries.md).

You can enable table scans to loosen the restrictions on the `WHERE` clause or eliminate 
the `WHERE` clause altogether. Table scans can be turned on for pull queries running in the 
current CLI session with the command `SET 'ksql.query.pull.table.scan.enabled'='true';`. 
They can also be enabled by default by setting a server configuration property with 
`ksql.query.pull.table.scan.enabled=true`.Once table scans are enabled, the following 
additional expressions are allowed:
-   Key column(s) using range comparisons to literals.
-   Non key columns to be used alone, without key references.
-   Columns to be compared to other columns.

> -   References to subsets of columns from a multi-column key.

Reviewer: Forgot to add this:
- Complex expressions without direct column references (e.g. `instr(NAME_COL, 'hello') > 0`)

Author: added

```sql
SELECT select_expr [, ...]
  FROM table
  [ WHERE where_condition ]
  [ AND window_bounds ];
```

Description
-----------

Pulls the current value from the materialized view and terminates. The result
of this statement is not persisted in a {{ site.ak }} topic and is printed out
only in the console. Because materialized views are incrementally updated as new 
events arrive, pull queries run with predictably low latency. They are a great match 
for request/response flows. For asynchronous application flows, see 
[Push Queries](select-push-query.md).

> You can execute a pull query by sending an HTTP request to the ksqlDB REST API, and

Reviewer: It streams back a chunked response.  Technically, it's a single response, though maybe it's a little misleading, since I think that wording was intended to mean it's returned in one fell swoop (one chunk!).  Maybe we want to change the description to either say "Streams back a chunked response as rows are fetched" or just make it a little more vague so that it's a little less tied to an implementation by saying "the API responds with resulting rows".

the API responds with a single response.  

-   Pull queries are expressed using a strict subset of ANSI SQL.
-   You can issue a pull query against any table that was created by using 
    a [CREATE TABLE AS SELECT](../../ksqldb-reference/create-table-as-select) 
    statement.

> -   `SELECT` statements can contain column arithmetic and function calls.

Author: ewmoved the commentary on `SELECT`. It was a bit confusing and wrong

-   Currently, we do not support pull queries against tables created by using a [CREATE TABLE](../../ksqldb-reference/create-table) statement.
-   `JOIN`, `PARTITION BY`, `GROUP BY` and `WINDOW` clauses aren't supported currently.

WHERE Clause Guidelines
-----------------------

By default, only key lookups are enabled. They have the following requirements:
-   Key column(s) must use an equality comparison to a literal (e.g. KEY = 'abc').
-   On windowed tables, WINDOWSTART and WINDOWEND can be optionally compared to literals. 
    For more information on windowed tables, see [Time and Windows in ksqlDB](../../concepts/time-and-windows-in-ksqldb-queries.md).

> You can enable table scans to loosen the restrictions on the `WHERE` clause or eliminate 

Reviewer 2: we might want to say what a "table scan" is (perhaps the note below is a good place for this). @JimGalasyn thoughts?

the `WHERE` clause altogether. Table scans can be turned on for pull queries running in the 
current CLI session with the command `SET 'ksql.query.pull.table.scan.enabled'='true';`. 
They can also be enabled by default by setting a server configuration property with 

> `ksql.query.pull.table.scan.enabled=true`.Once table scans are enabled, the following 

Reviewer 2: ```suggestion
`ksql.query.pull.table.scan.enabled=true`. Once table scans are enabled, the following 

```

additional expressions are allowed:
-   Key column(s) using range comparisons to literals.

> -   Non key columns to be used alone, without key references.

Reviewer 3: ```suggestion
-   Non-key columns to be used alone, without key references.

```

-   Columns to be compared to other columns.

> -   References to subsets of columns from a multi-column key.

Reviewer 2: do we also support UDFs and function calls? Might make sense to call out what is _not_ supported as opposed to what is here.

Reviewer: Yeah, we do support UDFs and function calls.  Really, anything goes.  The below line gives an example of this: `(e.g. instr(NAME_COL, 'hello') > 0)`

	greater performance and efficiency.

Examples
--------
Pull queries against a table `TOP_TEN_RANKS` created by using a 
[CREATE TABLE AS SELECT](../../ksqldb-reference/create-table-as-select) statement:
First, create a table named `GRADES` by using a [CREATE TABLE](../../ksqldb-reference/create-table) 
statement:
```sql
CREATE TABLE GRADES (ID INT PRIMARY KEY, GRADE STRING, RANK INT) 
  WITH (kafka_topic = 'test_topic', value_format = 'JSON', partitions = 4);
```
Then, create a derived table named `TOP_TEN_RANKS` by using a 
[CREATE TABLE AS SELECT](../../ksqldb-reference/create-table-as-select) statement:
```sql
CREATE TABLE TOP_TEN_RANKS 
  AS SELECT ID, RANK 
  FROM GRADES 
  WHERE RANK <= 10;
```
If you want to look up only the student with `ID = 5` in the `TOP_TEN_RANKS` table using a pull query:
```sql
SELECT * FROM TOP_TEN_RANKS
  WHERE ID = 5;
```
After enabling table scans, you can fetch the current state of your `TOP_TEN_RANKS` table using a pull query:
```sql
SELECT * FROM TOP_TEN_RANKS;
```
If you want to look up the students whose ranks lie in the range `(4, 8)`:
```sql
SELECT * FROM TOP_TEN_RANKS
  WHERE ID > 4 AND ID < 8;
```

> Pull queries against a table `INNER_JOIN` that is created by joining multiple tables:

Reviewer 2: I'm not entirely sure this example is helpful (the windowed one below is helpful because it shows the usage of window bounds on key lookups)

Author: We have stated in our 0.17 blog and 0.17 docs explicitly that Pull queries are not supported on Table-Table joins. This just clarifies to a certain degree that they are now supported.

Reviewer 2: IMO I think it's sufficiently clarified above when you discuss that any CTAS table can be queried. Not a big deal though, we can keep it or get another opinion on it.

Author: Let's keep it. It is also one of the examples in our 0.18 blog

the API responds with a single response.  

The WHERE clause must contain a value for each primary-key column to retrieve and may

> optionally include bounds on `WINDOWSTART` and `WINDOWEND` if the materialized view is windowed.

Reviewer: We should clarify this in more detail for when table scans are enabled.  We should say, after the "Time and WIndows" link, that enabling table scans removes many of the restrictions and allows most queries, though it'll be less performant, depending on the size of the underlying view.

For more information, see 
[Time and Windows in ksqlDB](../../concepts/time-and-windows-in-ksqldb-queries.md).

Example
-------
You can issue a pull query against a derived table that was created by using 
the [CREATE TABLE AS SELECT](../../ksqldb-reference/create-table-as-select) 

> statement. For example, you can create a table named `GRADES` by using a [CREATE TABLE](../../ksqldb-reference/create-table) 

Reviewer: I think we should have two sections: `Queryable views` and `WHERE clause guidelines`.  They could be before the examples and just summarize each.

`Queryable views` would state Almog's assertion like:
Any table created with `AS SELECT` can be queried.  Some examples include:
- Tables created by querying another table
- Tables created with a group by clause
- Tables created by joining other tables.

`WHERE clause guidelines` would state that by default, only key lookups are enabled.  These have the following requirements:
- Key column(s) must use an equality comparison to a literal (e.g. `KEY = 'abc'`)
- On windowed tables, `WINDOWSTART` and `WINDOWEND` can be optionally compared to literals.

To enable table scans, you set `ksql.query.pull.table.scan.enabled=true`.  Once they're are enabled, additional expressions are allowed:
- Key column(s) using range comparisons to literals
- Non key columns to be used alone, without key references
- Columns to be compared to other columns
- References to subsets of columns from a multi-column key

Optionally, these can be worked into the examples, but I think it would be useful to summarize.  Also should probably add some more examples that hit some of the above, especially table scans.

the API responds with a single response.  

The WHERE clause must contain a value for each primary-key column to retrieve and may
optionally include bounds on `WINDOWSTART` and `WINDOWEND` if the materialized view is windowed.
For more information, see 
[Time and Windows in ksqlDB](../../concepts/time-and-windows-in-ksqldb-queries.md).

Example
-------
You can issue a pull query against a derived table that was created by using 
the [CREATE TABLE AS SELECT](../../ksqldb-reference/create-table-as-select) 

> statement. First, create a table named `GRADES` by using a [CREATE TABLE](../../ksqldb-reference/create-table) 

Reviewer 3: ```suggestion
statement. For example, you can create a table named `GRADES` by using a [CREATE TABLE](../../ksqldb-reference/create-table) 

```

statement:
```sql
CREATE TABLE GRADES (ID INT PRIMARY KEY, GRADE STRING, RANK INT) 
  WITH (kafka_topic = 'test_topic', value_format = 'JSON', partitions = 1);
```
Then, create a derived table named `TOP_TEN_RANKS` by using a 
[CREATE TABLE AS SELECT](../../ksqldb-reference/create-table-as-select) statement:
```sql
CREATE TABLE TOP_TEN_RANKS 
  AS SELECT ID, RANK 
  FROM GRADES 
  WHERE RANK <= 10;
```

> You can fetch the current state of your materialized view `TOP_TEN_RANKS` by using a pull query:

Reviewer 3: ```suggestion
You can fetch the current state of the `TOP_TEN_RANKS` materialized view by using a pull query:

```

```sql
SELECT * FROM TOP_TEN_RANKS;
```
If you want to look up only the student with `ID = 5` in the materialized view:
```sql
SELECT * FROM TOP_TEN_RANKS
  WHERE ID = 5;
```

> If the materialized view `pageviews_by_region` is windowed:

Reviewer 3: ```suggestion
If the `pageviews_by_region` materialized view is windowed:

```

If no bounds are placed on `WINDOWSTART` or `WINDOWEND`, rows are returned for all windows
in the windowed table.

You can also issue pull queries against materialized views that are created using 
joining multiple tables:
```sql
CREATE TABLE LEFT_TABLE (ID BIGINT PRIMARY KEY, NAME varchar, VALUE bigint) 
  WITH (kafka_topic='left_topic', value_format='json', partitions=4);
```
```sql
CREATE TABLE RIGHT_TABLE (ID BIGINT PRIMARY KEY, F1 varchar, F2 bigint) 
  WITH (kafka_topic='right_topic', value_format='json', partitions=4);
```
```sql
CREATE TABLE INNER_JOIN AS SELECT L.ID, NAME, VALUE, F1, F2 FROM LEFT_TABLE L JOIN RIGHT_TABLE R ON L.ID = R.ID;
```

> You can fetch the current state of your materialized view `INNER_JOIN` by using a pull query:

Reviewer 3: ```suggestion
You can fetch the current state of the `INNER_JOIN` materialized view by using a pull query:

```
